### PR TITLE
Make Dimer Method N-Vector Translationally Invariant

### PR DESCRIPTION
**Pull request by bapride11**
Add code to check if N-vector is translationally invariant (sum N_x = sum N_y = sum N_z = 0) before each update, and shift by average component along direction if not.

### CODE REVIEW

File: pysisyphus/calculators/Dimer.py [hunk 1]
```diff
@@ -352,13 +353,31 @@ def set_N_raw(self, coords):
...
 
         self.log(f"Initial orientation:\n\t{self.N}")
 
+    def make_translationally_invariant(self, coords):
+        # Sum vector components over cartesian directions (x,y,z)
+        # Sum over each direction should equal zero if translationally invariant
+        sum = np.zeros(3)
+        for direction, component in enumerate(coords):
+            sum[direction % 3] += component
+        average = sum / (coords.size / 3)
```
Comment (eljost) [inline pysisyphus/calculators/Dimer.py:369]: The first loop can be avoided by using the built-in method `mean()` on a reshaped `coords` array along
the first axis (0).
```suggestion
        average = coords.reshape(-1, 3).mean(axis=0)
```

File: pysisyphus/calculators/Dimer.py [hunk 2]
```diff
@@ -352,13 +353,31 @@ def set_N_raw(self, coords):
...
             N_raw = np.sum(bond_modes, axis=0)
-        # Normalize N_raw
+        # Make N_raw translationally invariant and normalize
         self.N = N_raw
         # Now we keep the normalized dimer orientation
         self.N_raw = self.N
 
         self.log(f"Initial orientation:\n\t{self.N}")
 
+    def make_translationally_invariant(self, coords):
```
Comment (eljost) [inline pysisyphus/calculators/Dimer.py:363]: I would call the method something like `remove_translation()` and rename the `coords` argument to something more general like `arr` or `vec` or even `displacement`.

File: pysisyphus/calculators/Dimer.py [hunk 3]
```diff
@@ -352,13 +353,31 @@ def set_N_raw(self, coords):
...
+        average = sum / (coords.size / 3)
+        
+        if max(abs(average)) > 1e-8:
+            self.log(f"N-vector not translationally invariant. Shifting by ({average[0]}, {average[1]}, {average[2]}) before normalization.")
+        else:
+            return coords
+        # Subtract the average component along each direction to make sum zero
+        invariant_coords = np.copy(coords)
+        for direction in range(0, invariant_coords.size):
+            invariant_coords[direction] -= average[direction % 3]
```
Comment (eljost) [inline pysisyphus/calculators/Dimer.py:378]: Similarly, this loop can be avoided by using numpy array broadcasting.
```suggestion
            invariant_coords = (coords.reshape(-1, 3) - average[None, :]).flatten()
```

**Comment by eljost** [comment]
Hey Greg,
thanks for the PR! Even though I proposed some changes, this PR makes me happy, as it is the first one from outside our group.

Please add a argument/flag, controlling (enable/disable) the removal of translation `rotation_remove_trans`. Otherwise the Dimer method would not work with the built-in analytical potentials, which have only 3 coordinates. Making such an N vector translationally invariant would always result in a 0-vector. So it must be possible to disable the removal for such cases (see tests/test_dimer/test_dimer.py, test_dimer().

It would also nice to have a quick test for this feature that uses pyscf. Please have a look at `tests/test_baker_ts_dimer/test_baker_ts_dimer.py, test_baker_dimer_new().

Let me know if you need help with this.

**Comment by bapride11** [comment]
Hi Johannes!

Thank you for the suggestions! This is my first code contribution anywhere, and sometimes it's hard to break the C paradigm in my head.

For the tests, would you like it to do the whole Baker TS set, and if so, should we test with both providing initial N-vectors and utilizing the randomly generated N-vector?

Best,
Greg

**Comment by eljost** [comment]
> This is my first code contribution anywhere, and sometimes it's hard to break the C paradigm in my head.

No worries!

One (parametrized) test would be enough, that uses the flag `rotation_remove_trans` and pyscf.
```python
@using("pyscf")
@pytest.mark.parametrize(
    "rotation_remove_trans", [
        True,
        False,
    ]
)
def test_remove_translation(rotatition_remove_trans):
    ...
```
Currently the CI fails, because removal of translation breaks two Dimer tests using an analytical potential (only 1 atom).
```
FAILED tests/test_dimer/test_dimer.py::test_dimer[direct-9] - assert False
FAILED tests/test_dimer/test_dimer.py::test_dimer[fourier-9] - assert False
```

You can run tests in the tests subdirectory using pytest (`pytest --collect-only` to list available tests and `pytest -s -v [-k [one of the test names from --collect-only]`).

**Comment by eljost** [comment]
Thanks for the fixes. This looks complete now, isn't it? If so I'll merge it soon.

The performance for both test cases is similar, but this may also be related to the example, and the fact, that the guess for N_raw is already pretty reasonable from the beginning. But overall, it was already proven that this approach is useful (in combination with removal of rotation) as discussed [10.1021/ct501155k](https://pubs.acs.org/doi/abs/10.1021/ct501155k).

All the best,
Johannes

**Comment by eljost** [comment]
Yes, some directories are excluded, from starting a CI as seen [here](https://github.com/eljost/pysisyphus/blob/9acb41c0b657a0543813c1c674deb5ce025c2b88/.github/workflows/pythonapp.yml#L5), at least for pushes...

You can add the new keyword [here](https://github.com/eljost/pysisyphus/blob/master/docs/tsoptimization.rst)

**Comment by bapride11** [comment]
I've documented the new feature and this should be ready to merge.

**Comment by eljost** [comment]
Really nice, thanks for the work Greg.

**Pull request merged**